### PR TITLE
Query strings

### PR DESCRIPTION
**Pull request by seemantk**
This is the overall pull request which will include a number of changes:

1. Sorting widget
2. Persistent choices
  * When a year or sort order is selected, it should persist through section and question changes.
3. Direct URL navigation to focus on:
  * section
    * done
  * question
  * focus on year
    * highlight row on grid
    * show choropleth on map
  * focus on state
    * highlight cell on grid
    * highlight state on map

**URLs & Query Strings**
The URL components encapsulate all the controls that affect the data shown:

#section-name?question=database-key&year=2XXX&state=ST

section-name is validated with the tabs (click to navigate)
question is derived from the dropdowns
year is validated with the appropriate dropdown
state is validated with the appropriate dropdown.

_Behaviours_
If the URL contains:
- 'section-name' only
`Default to the first option in each dropdown of the appropriate tab, first option in year dropdown, no options in state dropdown`
- 'year' only
`Default to the first navigation tab, then follow 'section-name only', then select year in dropdown`
- 'question' only
`Tricky to derive section from this information.  Basically, find the <option> with that info, and then traverse through its parentNode hierarchy.`
- 'state' only
`follow 'section-name only' and highlight a state`

### CODE REVIEW

**Comment by curran** [comment]
@seemantk Nice!

The UI elements don't seem to be inside the map box. This is what I'm seeing:

image

**Comment by curran** [comment]
Boom! Fixed.

image

The issue was that we were using two nested `.column`s, without a `row` in between to break up the grid in the right way. I fixed the issue by inserting a `.row` div.

**Comment by curran** [comment]
Feeling inspired to work in this branch, hope you don't mind ;)

I tweaked the UI a bit:

image

 * Added label for year
 * Swapped locations (I feel like the right side is more "associated" with the grid than the left, which is more universal, appropriate for the year).
 * Used `.pull-right` to get the grid sorting widget to fill out the right side of the panel footer.

**Comment by curran** [comment]
@seemantk My thoughts on how we can achieve splitting the year selector and sorting mode:

 * We introduce a new dispatch type, `"sortMode"`.
 * My gut feeling is that we could get rid of this `msg` thing in [grid.js#L158](https://github.com/cfinst/cfinst.github.io/blob/d584c2d0047643f12e00d3a08832405cb8fa6739/js/grid.js#L158), and have the `Atlas` component be responsible for filtering its own data, with knowledge of the sorted year. It would be nice to have no coupling between the grid and the map, so the grid can re-render when the sorting mode changes, while the map is *not* re-rendered. Currently re-rendering of the grid *always* causes re-rendering of the map, which is not ideal.
 * It would be simpler to reason about sorting if we invocation of `resort()` from `grid.selectedYear()` at [grid.js#L464](https://github.com/cfinst/cfinst.github.io/blob/d584c2d0047643f12e00d3a08832405cb8fa6739/js/grid.js#L464), and always invoke `resort()` in the render function, modifying what's there at [grid.js#L54](https://github.com/cfinst/cfinst.github.io/blob/d584c2d0047643f12e00d3a08832405cb8fa6739/js/grid.js#L54). Instead of `if(sortYear) resort();`, this could be `resort(sortMode, selectedYear);`, the the internals of `resort` can determine if real re-sorting is required, by comparing to previous values passed in.

**Comment by curran** [comment]
@seemantk It looks like the breakage is due to this selection on [schematic.js#L273](https://github.com/cfinst/cfinst.github.io/blob/6e3e70a5b71d19ec8377619a749950323f9e1d92/js/schematic.js#L273) coming up empty

```js
d3.selectAll("#{{ section[0] }} select")
```

Any idea why this would be coming up empty?

**Comment by curran** [comment]
@seemantk I spent a good chunk of time trying to fix the breakage, investigating the part inside `{% comment %}  // CONTRIBUTION LIMITS CUSTOMIZATIONS` , but I have not able to break through. Blocked on this presently for pushing ahead with the sort mode work.

File: _data/sections/contribution-limits/legends.yml [hunk 1]
```diff
@@ -1,59 +1,64 @@
...
-      color: "#d1e5f0"
-    - min: 50001
-      max: 1000000
-      label: $50,000 or greater
-      color: "pink"
-    - min: 1000001
-      max: Infinity
-      label: Unlimited
-      color: "#cb181d"
+- name: default
```
Comment (seemantk) [inline _data/sections/contribution-limits/legends.yml]: Thanks!  The data files kept bugging me until I did this change last night.

**Comment by seemantk** [comment]
Sorry about that @curran - I'd hit a wall of fatigue around 1:30 am so went to bed.  You're right in that contrib-limits haven't been made to work yet -- the huge comment-blocked code needs to be examined and ported over into tabulus.

File: js/tabulus.js [hunk 2]
```diff
@@ -34,8 +34,9 @@ function Tabulus() {
                   })
             ;
         }
-        dropdown.each(function(d, i) {
-            // Call the "change" handler function for each dropdown
+        // Call the "change" handler function for the first dropdown to trigger render
+        container.select("select").each(function(d, i) {
```
Comment (curran) [inline js/tabulus.js]: Nice catch!

File: _data/sections/contribution-limits/controls.yml [hunk 3]
```diff
@@ -1,40 +1,40 @@
 - group: donor
   dropdowns:
-    - id: donor
+    - question: donor
       label: Donor
       options:
-        - text: Individual
+        - label: Individual
```
Comment (curran) [inline _data/sections/contribution-limits/controls.yml]: Nice consistency improvement.

File: js/schematic.js [hunk 4]
```diff
@@ -260,37 +260,3 @@
     })
   ;
 }());
-
-
-
-
-{% comment %}  // CONTRIBUTION LIMITS CUSTOMIZATIONS
```
Comment (curran) [inline js/schematic.js]: Great work porting this into Tabulus!

**Comment by seemantk** [comment]
@curran, thank you for updating UI/UX status on this branch.

Algorithmically, both types of widget sets (Q&A and Multi-dropdown) now do similar things with [an object called "query"](https://github.com/cfinst/cfinst.github.io/pull/161/files#diff-ed588b740bd7a189e3e3d35b611c5635R32).

These "query" objects currently reside within tabulus -- so each Tab on our viz has its own query object.

So, for the blank "branch" dropdown -- the "query" object seems like an appropriate place to store previous value when the branch gets disabled (and blanked out).

File: js/grid.js [hunk 5]
```diff
@@ -262,15 +261,6 @@ function Grid(){
       }
   } // domainify()
 
-  function score() {
```
Comment (curran) [inline js/grid.js]: Nice, yeah I wanted to remove this for a while now. Thanks!

File: js/schematic.js [hunk 6]
```diff
@@ -222,6 +223,11 @@
       ;
   } // setupTabNavigation()
 
+  /*
+  ** Set up the colorScale object, which is keyed by the tab/section
+  ** The colors and labels for the various colorScales are all defined in
+  ** Jekyll, so we're using Jekyll template statements to populate the object.
```
Comment (curran) [inline js/schematic.js]: Awesome.

File: js/schematic.js [hunk 7]
```diff
@@ -18,7 +18,9 @@
           .connect(signal)
     , tip = d3.tip().attr('class', 'd3-tip')
     , tabs = {}
-    , query = {}
+    , query = { // These are the defaults
```
Comment (curran) [inline js/schematic.js]: Nice way of setting up defaults.

File: js/schematic.js [hunk 8]
```diff
@@ -183,17 +184,20 @@
...
           if(arg[0].length && arg[1].length)
-              vars[arg[0]] = decodeURIComponent(arg[1]);
+              args[arg[0]] = decodeURIComponent(arg[1]);
         })
       ;
-      query.section = window.location.hash.substring(1).toLowerCase()
-        || query.section
-      ;
+      // Now populate the query object from the URL
+      d3.keys(query).forEach(function(k) { query[k] = args[k]; });
```
Comment (curran) [inline js/schematic.js]: I'm not seeing the case of no params in the initial URL handled anywhere yet. Maybe in a future commit..

File: js/schematic.js [hunk 9]
```diff
@@ -79,19 +79,20 @@
...
-
       // Initialize the selected year to the most recent.
       signal.call("selectYear", null, d3.select("#chooser-year").node().value);
 
-      d3.select("a[href='#" + query.section + "']")
-          .node()
-        .click()
-      ;
+      // If the query section doesn't have a valid link, default to the first tab
+      var sec = d3.select("a[href='#" + query.section + "']");
```
Comment (curran) [inline js/schematic.js]: Nice, there we go, I spoke too soon. The case of no params in URL is handled here, for section at least.

**Comment by curran** [comment]
@seemantk Perhaps in the future it would be best to keep things in a working state for feature work, and separate the refactoring work from the feature work (separate PRs for refactorings). I spent some time today trying to get the broken menus working again, and again was not able to break through to work on the year/sorting independence feature.

My gut feeling is that once things are in a broken state, it may be quite a heavy lift to restore the functionality again. I'm not sure if this current refactoring path will lead back to a working state soon, or not, so part of me wants to create a new branch from the current master, and just port over the UI changes from here to there and work on the year/sorting independence feature (separately from adding the URL parameters, and separate from Jekyll-related refactoring).

Leaving this for now, as my energy is spent and I'm feeling frustrated. Let me know how we might best move forward from here.

File: js/schematic.js [hunk 10]
```diff
@@ -276,9 +276,9 @@ navs["{{ section[0] }}"] = function(data) {
     ;
 
     // Set up the legend so it can be toggled depending on the donor.
-    tab.container(d3.select("#{{ section[0] }}"));
-
+    tabs["{{ section[0] }}"].colorScale(colorScale).grid(grid);
```
Comment (curran) [inline js/schematic.js]: Narrowed down the menu breakage to this commit. Branching from the commit before this.

**Comment by curran** [comment]
Narrowed down the menu breakage to this commit 6479f6494a30f430e5b7b60dca1d6ea891a2f411 . Branching from the commit before that one to work on year + sort mode isolation.

**Comment by seemantk** [comment]
Well, the slog continues, but now URL query strings are starting work.
Query strings are written out into the history accurately consistently.
Query strings are mostly being read into view accurately as well, except for recipient branch, which always seems to default to 'H'

I've begun porting @curran's logic from grid over, to figure out the color lookups.

But it's pi o'clock in the morning for me, and I'm sleepy.

Summary:

I'm overall happy with the progress, as I had to basically disconnect all the functionality and restore one by one.
Grid, dropdowns, and map all update.  Highlighting is broken still and coloring algorithm is broken still.

**Comment by seemantk** [comment]
Will rebase to master in the morning and reconnect the highlight functionalities.

**Comment by seemantk** [comment]
Well, it doesn't work after that merge, so I have some debugging to do tomorrow.  See you on the call.

**Comment by seemantk** [comment]
Slowly regaining functionality after the merge.  Will continue this tonight.

currentstatus

**Comment by seemantk** [comment]
to do this evening:

- reconnect the year and sorter signals
- figure out why going from contrib-limits to other-restrictions doesn't refresh the default legend
- verify the keycolumn stuff in atlas.

**Comment by curran** [comment]
@seemantk Great stuff! Looking forward to seeing fully restored functionality.

**Comment by seemantk** [comment]
@curran I think this is basically ready for your review.

Meanwhile, I think I've gotten URL queries (case-sensitive) working as intended, and I've reconnected the functionalities for state highlighting and the mouseover interactions.

Things that I know of that are still broken (and I need help/fresh pair of eyes):

- sort-by-value in the grid.  the actual .sort() call doesn't seem to get called.  I can't figure out why.

Would appreciate your help, guidance, review and input at this time :)

**Comment by seemantk** [comment]
@curran oh yeah, I changed the stroke-dasharray of the yellow rect in the grid to match the stroke-dasharray  of the matching state.  this is because the state that is selected (via clicking on a rect or a state or by URL query) is in solid stroke.

I also couldn't figure out how to add a rect to the grid's overlay for the clicked state (to match query.state and query.year).

**Comment by seemantk** [comment]
I've noticed an inconsistency in the dropdown enabling/disabling behaviour (for recipient branch).  Addressing, and will update the comment.

**Comment by curran** [comment]
Thanks @seemantk ! Having a look now, these are the things I notice:

 * When the page loads, I see that the "Branch" dropdown selection is empty:
image
 * I'm not sure I like the dashed lines. Maybe we could use a different color, or a different thickness, to differentiate between hovered and selected (love the new concept of selection!).
 * The selected state is being stroked with solid yellow on the map, but there is no corresponding selection highlight on the grid.
 * The year is no longer being bolded.
 * After selecting to sort "By Value" then changing the year, the grid content disappears.
 * "Download Visible Data" button is not working. Perhaps my fix from earlier was not merged into here?
 * When switching from one tab to another, the question dropdown selection becomes blank.
 * When switching to the contribution limits tab, the "Donor" and "Branch" dropdown selections become blank.

**Comment by seemantk** [comment]
Curran, thank you for reviewing the changes so far.

1.  The empty dropdowns are as a result of some work I was doing last night to blank out a disabled dropdown.  I'll push a fix today.
2. +1 to rethinking the colors/thicknesses/outline styles.
3. I was hoping to punt this one to you, actually ;)
4. Will push fix for this, shortly 
5. Yes, this one confounds me ([comment from yesterday](https://github.com/cfinst/cfinst.github.io/pull/161#issuecomment-285930292))

**Comment by seemantk** [comment]
Starting to address these:

- Empty dropdowns issue now addressed.
  - Disabled/enabled values is still buggy
- The dropdown values are remembered when the tabs change, and restored upon return.

**Comment by curran** [comment]
@seemantk Great! Will have a look.

**Comment by seemantk** [comment]
Another step to URL query + dropdown inits.  The branch value is retained through disables (and even through tab changes).  Verify by selecting "Candidate" again.

The bug is visual only: the disabled value is the first option.

**Comment by curran** [comment]
@seemantk Nice improvements to the defaults behavior.

I just read through the overall diff for this PR. My perception with this PR is that there's a lot of nice refactoring work that's been done, but it's also coupled with the half-implemented new feature of selected states, the new feature of the URL parameters, and there are some feature regressions that also come along with it (the grid disappearing).

Probably I could dive in and fix the disappearing grid issue, but IDK it feels like a quite time consuming task, navigating all the new refactorings and trying to bug fix, especially given that I already encountered and fixed the disappearing grid issue once in #163.

I recall you mentioned the analogy of a bicycle pedaling, and we work together best when it's like we're each responsible for pedaling one side, then the other can take over. I feel like this PR kind of goes against that, where it's like backpedaling and ends up being more work than required.

I'm not sure what the best way forward here is. I'd like to ultimately see your work here land into the project, but there are some things standing in the way of a merge:

 * The half-working state selection feature is not delivarable as-is. I'd say it might be best to remove / comment-out this feature for now, restoring the original styling for hover highlights, and circle back to add the feature in a future PR. Would you mind making this change?
 * The disappearing grid is blocking merge. I can spend time looking into this, comparing with how I fixed this before in #163.

**Comment by curran** [comment]
@seemantk Just pushed a commit that fixes the disappearing grid. Looks like it was to do with number parsing.

**Comment by curran** [comment]
Poking around in the menus, everything looks great! :+1: 

The only thing blocking merge at this point is the state outlines, and the partially working state selection feature. @seemantk How do you feel about reverting this part back to how it was before, and tacking the feature in a future PR?

**Comment by curran** [comment]
Oh one small thing - the red map with Party -> Party is really cool, but I believe it should be all marked as "Prohibited", rather than "Unlimited", is that right? Or maybe "Unimited" is the correct value to have here? I don't remember the specifics of the discussion we had with @bglavin on this direction. There were sometimes "transfers" from party to party, but no "donations", so I'm not sure..

image

**Comment by seemantk** [comment]
OK, that visual bug is gone now, as well.

**Comment by seemantk** [comment]
@curran thanks for the sorting fix!

I agree - I'll put the state outline stuff into a separate PR

**Comment by curran** [comment]
@seemantk Awesome :smile: 

If you're wrapping up for the day, I'd be happy to take over from here and disable the state outline stuff. Just let me know, maybe point towards where the changes need to happen (probably the URL handling part will need to change, as well as the interaction bits).

**Comment by curran** [comment]
Just about ready for merge. Just noticed that the download button is not working.

**Comment by curran** [comment]
All set. Verified locally all the various features and cases. Will merge.

**Pull request merged**